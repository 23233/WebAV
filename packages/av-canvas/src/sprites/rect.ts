import { IPoint } from '../types'

export type TCtrlKey = 't' | 'b' | 'l' | 'r' | 'lt' | 'lb' | 'rt' | 'rb' | 'rotate'

export const CTRL_KEYS = ['t', 'b', 'l', 'r', 'lt', 'lb', 'rt', 'rb', 'rotate']

interface IRectBaseProps {
  x: number
  y: number
  w: number
  h: number
  angle: number
}

type TKeyFrameOpts = Record<`${number}%` | 'from' | 'to', Partial<IRectBaseProps>>

interface IAnimationOpts {
  duration: number
  delay?: number
  iterationCount?: number
}

export type TAnimationKeyFrame = Array<[number, Partial<IRectBaseProps>]>

export class Rect implements IRectBaseProps {
  /**
   * ctrl 节点的边长
   */
  static CTRL_SIZE = 16

  x = 0
  y = 0
  w = 0
  h = 0
  angle = 0

  /**
   * ctrl.master is sprite rect
   */
  master: Rect | null = null

  #animatKeyFrame: TAnimationKeyFrame | null = null

  #animatOpts: Required<IAnimationOpts> | null = null

  constructor (x?: number, y?: number, w?: number, h?: number, master?: Rect | null) {
    this.x = x ?? 0
    this.y = y ?? 0
    this.w = w ?? 0
    this.h = h ?? 0
    this.master = master ?? null
  }

  get center (): IPoint {
    const { x, y, w, h } = this
    return { x: x + w / 2, y: y + h / 2 }
  }

  // 上下左右+四个角+旋转控制点
  get ctrls (): Record<TCtrlKey, Rect> {
    const { w, h } = this
    // todo：控制点在高分辨率下看起来太小
    // 控制点元素大小, 以 分辨率 为基准
    const sz = Rect.CTRL_SIZE
    // half size
    const hfSz = sz / 2
    const hfW = w / 2
    const hfH = h / 2
    // rotate size
    const rtSz = sz * 1.5
    const hfRtSz = rtSz / 2
    // ctrl 坐标是相对于 sprite 中心点
    return {
      t: new Rect(-hfSz, -hfH - hfSz, sz, sz, this),
      b: new Rect(-hfSz, hfH - hfSz, sz, sz, this),
      l: new Rect(-hfW - hfSz, -hfSz, sz, sz, this),
      r: new Rect(hfW - hfSz, -hfSz, sz, sz, this),
      lt: new Rect(-hfW - hfSz, -hfH - hfSz, sz, sz, this),
      lb: new Rect(-hfW - hfSz, hfH - hfSz, sz, sz, this),
      rt: new Rect(hfW - hfSz, -hfH - hfSz, sz, sz, this),
      rb: new Rect(hfW - hfSz, hfH - hfSz, sz, sz, this),
      rotate: new Rect(-hfRtSz, -hfH - sz * 2 - hfRtSz, rtSz, rtSz, this)
    }
  }

  clone (): Rect {
    const { x, y, w, h, master } = this
    return new Rect(x, y, w, h, master)
  }

  setAnimation (keyFrame: TKeyFrameOpts, opts: IAnimationOpts): void {
    this.#animatKeyFrame = Object.entries(keyFrame)
      .map(([k, val]) => {
        const numK = ({ from: 0, to: 100 })[k] ?? Number(k.slice(0, -1))
        if (isNaN(numK) || numK > 100 || numK < 0) {
          throw Error('keyFrame must between 0~100')
        }
        return [numK / 100, val]
      })
    this.#animatOpts = Object.assign({
      duration: opts.duration * 1e6,
      delay: (opts.delay ?? 0) * 1e6,
      iterationCount: opts.iterationCount ?? Infinity
    })
  }

  animate (time: number): void {
    if (
      this.#animatKeyFrame == null ||
      this.#animatOpts == null ||
      time < this.#animatOpts.delay
    ) return
    // todo: delay, other timing-function
    Object.assign(
      this,
      linearTimeFn(time, this.#animatKeyFrame, this.#animatOpts)
    )
  }

  /**
   * 检测点击是否命中
   */
  checkHit (tx: number, ty: number): boolean {
    let { angle, center, x, y, w, h, master } = this
    // ctrls 的中心点、旋转角度都去自于 master （sprite）
    const cnt = master?.center ?? center
    const agl = master?.angle ?? angle
    // ctrl 初始化时其坐标就是相对于 master 的，参见 get ctrls()
    // 所以此处不用转换
    if (master == null) {
      x = x - cnt.x
      y = y - cnt.y
    }
    // 鼠标点击坐标映射成以中点为原点的坐标
    const tOX = tx - cnt.x
    const tOY = ty - cnt.y
    // 如果有旋转，映射成相对 sprite 原点，旋转前的坐标
    let mx = tOX
    let my = tOY
    if (agl !== 0) {
      // 推导公式 https://github.com/hughfenghen/hughfenghen.github.io/issues/96
      mx = tOX * Math.cos(agl) + tOY * Math.sin(agl)
      my = tOY * Math.cos(agl) - tOX * Math.sin(agl)
    }

    if (mx < x || mx > x + w || my < y || my > y + h) return false

    return true
  }
}

export function linearTimeFn (
  time: number,
  kf: TAnimationKeyFrame,
  opts: Required<IAnimationOpts>
): Partial<IRectBaseProps> {
  if (time / opts.duration >= opts.iterationCount) return {}

  const t = time % opts.duration

  const process = time === opts.duration ? 1 : t / opts.duration
  const idx = kf.findIndex(it => it[0] >= process)
  if (idx === -1) return {}

  const startState = kf[idx - 1]
  const nextState = kf[idx]
  const startFrame = startState[1]
  const nextFrame = nextState[1]
  if (startFrame == null) return nextFrame

  const rs: Partial<IRectBaseProps> = {}
  // 介于两个Frame状态间的进度
  const stateProcess = (process - startState[0]) / (nextState[0] - startState[0])
  for (const prop in nextFrame) {
    const p = prop as keyof IRectBaseProps
    if (startFrame[p] == null) continue
    // @ts-expect-error
    // eslint-disable-next-line
    rs[p] = (nextFrame[p] - startFrame[p]) * stateProcess + startFrame[p]
  }

  return rs
}
