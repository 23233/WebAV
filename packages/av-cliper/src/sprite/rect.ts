import { EventTool } from '../event-tool';

interface IPoint {
  x: number;
  y: number;
}

export type TCtrlKey =
  | 't'
  | 'b'
  | 'l'
  | 'r'
  | 'lt'
  | 'lb'
  | 'rt'
  | 'rb'
  | 'rotate';

export const CTRL_KEYS = [
  't',
  'b',
  'l',
  'r',
  'lt',
  'lb',
  'rt',
  'rb',
  'rotate',
] as const;

export interface IRectBaseProps {
  x: number;
  y: number;
  w: number;
  h: number;
  angle: number;
}

export class Rect implements IRectBaseProps {
  /**
   * ctrl 节点的边长
   */
  static CTRL_SIZE = 16;

  #evtTool = new EventTool<{
    propsChange: (props: Partial<IRectBaseProps>) => void;
  }>();
  on = this.#evtTool.on;

  #x = 0;
  get x() {
    return this.#x;
  }
  set x(v) {
    this.#setBaseProps('x', v);
  }
  #y = 0;
  get y() {
    return this.#y;
  }
  set y(v) {
    this.#setBaseProps('y', v);
  }
  #w = 0;
  get w() {
    return this.#w;
  }
  set w(v) {
    this.#setBaseProps('w', v);
  }
  #h = 0;
  get h() {
    return this.#h;
  }
  set h(v) {
    this.#setBaseProps('h', v);
  }
  #angle = 0;
  get angle() {
    return this.#angle;
  }
  set angle(v) {
    this.#setBaseProps('angle', v);
  }

  #setBaseProps(prop: keyof IRectBaseProps, v: number) {
    const changed = this[prop] !== v;
    switch (prop) {
      case 'x':
        this.#x = v;
        break;
      case 'y':
        this.#y = v;
        break;
      case 'w':
        this.#w = v;
        break;
      case 'h':
        this.#h = v;
        break;
      case 'angle':
        this.#angle = v;
        break;
    }
    if (changed) this.#evtTool.emit('propsChange', { [prop]: v });

    // concat emit
    // if (changed) this.#nextTickEmitData[prop] = v;
    // Promise.resolve().then(() => {
    //   if (Object.keys(this.#nextTickEmitData).length > 0) {
    //     const data = this.#nextTickEmitData;
    //     this.#nextTickEmitData = {};
    //     this.#evtTool.emit('propsChange', { ...data });
    //   }
    // });
  }

  /**
   * ctrl.master is sprite rect
   */
  master: Rect | null = null;

  constructor(
    x?: number,
    y?: number,
    w?: number,
    h?: number,
    master?: Rect | null,
  ) {
    this.x = x ?? 0;
    this.y = y ?? 0;
    this.w = w ?? 0;
    this.h = h ?? 0;
    this.master = master ?? null;
  }

  get center(): IPoint {
    const { x, y, w, h } = this;
    return { x: x + w / 2, y: y + h / 2 };
  }

  fixedAspectRatio = false;

  // 上下左右+四个角+旋转控制点
  get ctrls() {
    const { w, h } = this;
    // 控制点元素大小, 以 分辨率 为基准
    const sz = Rect.CTRL_SIZE;
    // half size
    const hfSz = sz / 2;
    const hfW = w / 2;
    const hfH = h / 2;
    // rotate size
    const rtSz = sz * 1.5;
    const hfRtSz = rtSz / 2;
    // ctrl 坐标是相对于 sprite 中心点
    const tblr = this.fixedAspectRatio
      ? {}
      : {
          t: new Rect(-hfSz, -hfH - hfSz, sz, sz, this),
          b: new Rect(-hfSz, hfH - hfSz, sz, sz, this),
          l: new Rect(-hfW - hfSz, -hfSz, sz, sz, this),
          r: new Rect(hfW - hfSz, -hfSz, sz, sz, this),
        };
    return {
      ...tblr,
      lt: new Rect(-hfW - hfSz, -hfH - hfSz, sz, sz, this),
      lb: new Rect(-hfW - hfSz, hfH - hfSz, sz, sz, this),
      rt: new Rect(hfW - hfSz, -hfH - hfSz, sz, sz, this),
      rb: new Rect(hfW - hfSz, hfH - hfSz, sz, sz, this),
      rotate: new Rect(-hfRtSz, -hfH - sz * 2 - hfRtSz, rtSz, rtSz, this),
    };
  }

  clone(): Rect {
    const { x, y, w, h, master } = this;
    const rect = new Rect(x, y, w, h, master);
    rect.angle = this.angle;
    rect.fixedAspectRatio = this.fixedAspectRatio;
    return rect;
  }

  /**
   * 检测点击是否命中
   */
  checkHit(tx: number, ty: number): boolean {
    let { angle, center, x, y, w, h, master } = this;
    // ctrls 的中心点、旋转角度都去自于 master （sprite）
    const cnt = master?.center ?? center;
    const agl = master?.angle ?? angle;
    // ctrl 初始化时其坐标就是相对于 master 的，参见 get ctrls()
    // 所以此处不用转换
    if (master == null) {
      x = x - cnt.x;
      y = y - cnt.y;
    }
    // 鼠标点击坐标映射成以中点为原点的坐标
    const tOX = tx - cnt.x;
    const tOY = ty - cnt.y;
    // 如果有旋转，映射成相对 sprite 原点，旋转前的坐标
    let mx = tOX;
    let my = tOY;
    if (agl !== 0) {
      // 推导公式 https://github.com/hughfenghen/hughfenghen.github.io/issues/96
      mx = tOX * Math.cos(agl) + tOY * Math.sin(agl);
      my = tOY * Math.cos(agl) - tOX * Math.sin(agl);
    }

    if (mx < x || mx > x + w || my < y || my > y + h) return false;

    return true;
  }
}
