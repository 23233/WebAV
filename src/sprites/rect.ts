import { IPoint } from '../types'

export type TCtrlKey = 't' | 'b' | 'l' | 'r' | 'lt' | 'lb' | 'rt' | 'rb' | 'rotate'

export class Rect {
  x = 0
  y = 0
  w = 0
  h = 0
  angle = 0

  constructor (x?: number, y?: number, w?: number, h?: number) {
    this.x = x ?? 0
    this.y = y ?? 0
    this.w = w ?? 0
    this.h = h ?? 0
  }

  get center (): IPoint {
    const { x, y, w, h } = this
    return { x: x + w / 2, y: y + h / 2 }
  }

  // 上下左右+四个角+旋转控制点
  get ctrls (): Record<TCtrlKey, Rect> {
    const { w, h } = this
    // todo：控制点在高分辨率下看起来太小
    // 控制点元素大小, 以 分辨率 为基准
    const sz = 8
    const hfSz = sz / 2
    const hfW = w / 2
    const hfH = h / 2
    const rtSz = sz * 3
    const hfRtSz = rtSz / 2
    // 相对于中心点的位置
    return {
      t: new Rect(-hfSz, -hfH - hfSz, sz, sz),
      b: new Rect(-hfSz, hfH - hfSz, sz, sz),
      l: new Rect(-hfW - hfSz, -hfSz, sz, sz),
      r: new Rect(hfW - hfSz, -hfSz, sz, sz),
      lt: new Rect(-hfW - hfSz, -hfH - hfSz, sz, sz),
      lb: new Rect(-hfW - hfSz, hfH - hfSz, sz, sz),
      rt: new Rect(hfW - hfSz, -hfH - hfSz, sz, sz),
      rb: new Rect(hfW - hfSz, hfH - hfSz, sz, sz),
      rotate: new Rect(-hfRtSz, -hfH - sz * 3 - hfRtSz, rtSz, rtSz)
    }
  }

  clone (): Rect {
    const r = new Rect()
    r.x = this.x
    r.y = this.y
    r.w = this.w
    r.h = this.h
    return r
  }

  /**
   * 检测点击是否命中
   */
  checkHit (tx: number, ty: number): boolean {
    let { angle, center, x, y, w, h } = this
    // 鼠标点击坐标映射成 canvas坐标, 然后转换为以中点为原点的坐标
    const tOX = tx - center.x
    const tOY = ty - center.y
    x = x - center.x
    y = y - center.y
    // 如果有旋转，映射成相对 sprite 原点，旋转前的坐标
    let mx = tOX
    let my = tOY
    if (angle !== 0) {
    // 推导公式 https://github.com/hughfenghen/hughfenghen.github.io/issues/96
      mx = tOX * Math.cos(angle) + tOY * Math.sin(angle)
      my = tOY * Math.cos(angle) - tOX * Math.sin(angle)
    }

    if (mx < x || mx > x + w || my < y || my > y + h) return false

    return true
  }
}
