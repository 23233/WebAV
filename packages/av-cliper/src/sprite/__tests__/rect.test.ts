import { expect, test } from 'vitest';
import { Rect } from '../rect';

test('center', () => {
  const rect = new Rect(0, 0, 100, 100);
  expect(rect.center).toEqual({ x: 50, y: 50 });
});

test('ctrls', () => {
  const rect = new Rect(0, 0, 100, 100);
  expect(rect.ctrls).toMatchSnapshot();
});

test('clone', () => {
  const { x, y, w, h } = new Rect(0, 0, 100, 100).clone();
  expect([x, y, w, h]).toEqual([0, 0, 100, 100]);
  expect(new Rect(0, 0, 100, 100).clone()).toMatchSnapshot();
});

test('checkHit', () => {
  const rect = new Rect(100, 100, 100, 100);
  rect.x = 100;
  rect.y = 100;
  rect.w = 100;
  rect.h = 100;
  // 边界检查
  expect(rect.checkHit(99, 99)).toBe(false);
  expect(rect.checkHit(100, 100)).toBe(true);
  expect(rect.checkHit(200, 200)).toBe(true);
  expect(rect.checkHit(201, 201)).toBe(false);

  expect(rect.checkHit(150, 90)).toBe(false);
  rect.angle = Math.PI / 4;
  // 原位置（左上角顶点）不在 pos（正方形）旋转 45° 之后的范围内
  expect(rect.checkHit(100, 100)).toBe(false);
  // 旋转后正上方外移一点点的位置被覆盖进来了
  expect(rect.checkHit(150, 90)).toBe(true);
});

// 固定比例后，ctrls 将移除 t,b,l,r 控制点
test('fixedAspectRatio', () => {
  const rect = new Rect(0, 0, 100, 100);
  expect(Object.keys(rect.ctrls)).toEqual([
    't',
    'b',
    'l',
    'r',
    'lt',
    'lb',
    'rt',
    'rb',
    'rotate',
  ]);
  rect.fixedAspectRatio = true;
  expect(Object.keys(rect.ctrls)).toEqual(['lt', 'lb', 'rt', 'rb', 'rotate']);
});
