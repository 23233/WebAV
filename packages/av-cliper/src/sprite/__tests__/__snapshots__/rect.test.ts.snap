// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Rect > ctrls 1`] = `
{
  "b": Rect {
    "master": Rect {
      "master": null,
      "on": [Function],
    },
    "on": [Function],
  },
  "l": Rect {
    "master": Rect {
      "master": null,
      "on": [Function],
    },
    "on": [Function],
  },
  "lb": Rect {
    "master": Rect {
      "master": null,
      "on": [Function],
    },
    "on": [Function],
  },
  "lt": Rect {
    "master": Rect {
      "master": null,
      "on": [Function],
    },
    "on": [Function],
  },
  "r": Rect {
    "master": Rect {
      "master": null,
      "on": [Function],
    },
    "on": [Function],
  },
  "rb": Rect {
    "master": Rect {
      "master": null,
      "on": [Function],
    },
    "on": [Function],
  },
  "rotate": Rect {
    "master": Rect {
      "master": null,
      "on": [Function],
    },
    "on": [Function],
  },
  "rt": Rect {
    "master": Rect {
      "master": null,
      "on": [Function],
    },
    "on": [Function],
  },
  "t": Rect {
    "master": Rect {
      "master": null,
      "on": [Function],
    },
    "on": [Function],
  },
}
`;

exports[`ctrls 1`] = `
{
  "b": Rect {
    "fixedAspectRatio": false,
    "master": Rect {
      "fixedAspectRatio": false,
      "master": null,
      "on": [Function],
    },
    "on": [Function],
  },
  "l": Rect {
    "fixedAspectRatio": false,
    "master": Rect {
      "fixedAspectRatio": false,
      "master": null,
      "on": [Function],
    },
    "on": [Function],
  },
  "lb": Rect {
    "fixedAspectRatio": false,
    "master": Rect {
      "fixedAspectRatio": false,
      "master": null,
      "on": [Function],
    },
    "on": [Function],
  },
  "lt": Rect {
    "fixedAspectRatio": false,
    "master": Rect {
      "fixedAspectRatio": false,
      "master": null,
      "on": [Function],
    },
    "on": [Function],
  },
  "r": Rect {
    "fixedAspectRatio": false,
    "master": Rect {
      "fixedAspectRatio": false,
      "master": null,
      "on": [Function],
    },
    "on": [Function],
  },
  "rb": Rect {
    "fixedAspectRatio": false,
    "master": Rect {
      "fixedAspectRatio": false,
      "master": null,
      "on": [Function],
    },
    "on": [Function],
  },
  "rotate": Rect {
    "fixedAspectRatio": false,
    "master": Rect {
      "fixedAspectRatio": false,
      "master": null,
      "on": [Function],
    },
    "on": [Function],
  },
  "rt": Rect {
    "fixedAspectRatio": false,
    "master": Rect {
      "fixedAspectRatio": false,
      "master": null,
      "on": [Function],
    },
    "on": [Function],
  },
  "t": Rect {
    "fixedAspectRatio": false,
    "master": Rect {
      "fixedAspectRatio": false,
      "master": null,
      "on": [Function],
    },
    "on": [Function],
  },
}
`;
